athena-scripts/raw_stage_optimisation_solution/scripts/core_preprocessing_functions.py

```diff
@@ -167,15 +167,26 @@ def generate_raw_select_filter(json_data, database, table, filter_processed_dt):
             raise ValueError("filter value for event_processing_testing_criteria is not found within config rules")
         print(f'config rule: event_processing_testing_criteria | filter: {event_processing_testing_criteria_filter}')
 
-        sql=f'''select * from \"{database}\".\"{table}\"'''
-
+        if stage_table_exists:
+            sql=f'''select * from \"{database}\".\"{table}\" where_clause'''
+        else:  
+            subquery = f'''select *,
+			                row_number() over (
+				            partition by event_id
+				            order by cast(concat(year, month, day) as int) desc
+			            ) as row_num
+               		    from \"{database}\".\"{table}\" as t
+                        where_clause'''
+            sql=f'''select * from ({subquery}) where row_num = 1'''
+                    
         if event_processing_testing_criteria_enabled and event_processing_testing_criteria_filter is not None:
-            sql = sql + f' where {event_processing_testing_criteria_filter}'
+            sql = sql.replace('where_clause',  f' where {event_processing_testing_criteria_filter} ')
+    
         elif event_processing_selection_criteria_filter is not None:
             update_process_dt = event_processing_selection_criteria_filter.replace('processed_dt', str(filter_processed_dt))
-            sql = sql + f' where {update_process_dt}'
+            sql = sql.replace('where_clause', f' where {update_process_dt} ')
             if event_processing_selection_criteria_limit is not None and event_processing_selection_criteria_limit > 0:
-                sql = sql + f' limit {event_processing_selection_criteria_limit}'
+                sql = sql + f' limit {event_processing_selection_criteria_limit} '
 
```

Comment:
i don't really understand what this code is doing but what if it goes into the implicit `else` block, then `sql` will still contain 'where_clause'? i think the old concat using `+` avoids this issue as something is either appended or not, instead of 'where_clause' being left in the query